internal/auth/token.go

```diff
@@ -0,0 +1,190 @@
+/*
+   Copyright 2024 Docker Terraform Provider authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package auth
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// TokenProvider provides a valid authentication token for API requests
+type TokenProvider interface {
```

Comment:
question: Why do we need an interface here?

Reply:
moved to where it's used

Reply:
I was more thinking about, why do we _need_ an interface?

Reply:
Not sure i understand the question. Provider configures a TokenProvider.   That interface is implemented by multiple different structs, depending on what tokens are available or what the user configured. what would we do instead?